config/v1/types_cluster_version.go

```diff
@@ -215,6 +222,58 @@ type UpdateHistory struct {
 // ClusterID is string RFC4122 uuid.
 type ClusterID string
 
+// ClusterVersionCapabilitiesDefault is a constant representing the
+// default policy for optional, core cluster components.
+type ClusterVersionCapabilitiesDefault string
+
+const (
+	// IncludeClusterVersionCapabilities indicates that optional,
+	// core cluster components should be included by default.
+	IncludeClusterVersionCapabilities = "Include"
+
+	// ExcludeClusterVersionCapabilities indicates that optional,
+	// core cluster components should be excluded by default.
+	ExcludeClusterVersionCapabilities = "Exclude"
+)
+
+// ClusterVersionCapabilitiesSpec configures the installation of
+// optional, core cluster components.
+// +k8s:deepcopy-gen=true
+type ClusterVersionCapabilitiesSpec struct {
+	// inclusionDefault declares the default policy for optional,
+	// core cluster components.
+
+	// +optional
+	InclusionDefault ClusterVersionCapabilitiesDefault `json:"inclusionDefault,omitempty"`
+
+	// include declares capabilities that should be included,
+	// regardless of inclusionDefault.
+	// +listType=atomic
+	// +optional
+	Include []string `json:"include,omitempty"`
+
+	// exclude declares capabilities that should be excluded,
+	// regardless of inclusionDefault.
+	// +listType=atomic
+	// +optional
+	Exclude []string `json:"exclude,omitempty"`
+}
+
+// ClusterVersionCapabilitiesStatus describes the state of optional,
+// core cluster components.
+// +k8s:deepcopy-gen=true
+type ClusterVersionCapabilitiesStatus struct {
```

Comment:
Given a list of "all the capabilities" and a list of "enabled capabilities", is information missing from status?

Reply:
`A` is the set of all caps.
`I` is the set of included caps.
`E` is the set of excluded caps.
Each cap must be either included or excluded, so `I` and `E` are disjoint, and the union of `I` and `E` is `A`.

So you can take any two of those three sets and construct the one you're missing:

* `A = I ∪ E`
* `E = A - I`
* `I = A - E` 

If we have to pick two to include in `status`, picking `I` and `E` gives us all the data we need, and saves a few bytes by excluding the largest, which is `A`.

Reply:
I prefer to see `KnownCapabilities` and `EnabledCapabilities`.

Reply:
224e90ed pivots to `enabledCapabilities` and `knownCapabilities`.